Document changes to rclcpp's logging macros

audrow opened the pull request:
This PR documents the changes from ros2/rclcpp#1442.

clalancette commented on source/Releases/Release-Galactic-Geochelone.rst:
```diff
@@ -102,6 +102,12 @@ Waitable API was modified to avoid issues with the ``MultiThreadedExecutor``.
 
+Change in ``rclcpp``'s logging macros
+"""""""""""""""""""""""""""""""""""""
+Previously, the logging macros were vulnerable to a `format string attack <https://owasp.org/www-community/attacks/Format_string_attack>`_, where the format string is evaluated and can potentially execute code, read the stack, or cause a segmentation fault in the running program. To address this security issue, the logging macro accepts only string literals for it's format string argument.
```
```suggestion
Previously, the logging macros were vulnerable to a `format string attack <https://owasp.org/www-community/attacks/Format_string_attack>`_, where the format string is evaluated and can potentially execute code, read the stack, or cause a segmentation fault in the running program.
To address this security issue, the logging macro now accepts only string literals for it's format string argument.

If you previously had code like:

.. code-block::

  const char *my_const_char_string format = "Foo";
  RCLPP_DEBUG(get_logger(), my_const_char_string);

you should now replace it with:

.. code-block::

  const char *my_const_char_string format = "Foo";
  RCLCPP_DEBUG(get_logger(), "%s", my_const_char_string);
```

clalancette commented on source/Releases/Release-Galactic-Geochelone.rst:
```diff
@@ -102,6 +102,12 @@ Waitable API was modified to avoid issues with the ``MultiThreadedExecutor``.
+"""""""""""""""""""""""""""""""""""""
+Previously, the logging macros were vulnerable to a `format string attack <https://owasp.org/www-community/attacks/Format_string_attack>`_, where the format string is evaluated and can potentially execute code, read the stack, or cause a segmentation fault in the running program. To address this security issue, the logging macro accepts only string literals for it's format string argument.
+
+This change removes some convenience from the logging macros, as ``std::strings`` are no longer accepted as the format argument. ``std::stringstream`` types are still accepted as arguments to the stream logging macros. See `ros2/rclcpp#1442 <https://github.com/ros2/rclcpp/pull/1442>`_ for more details.
```
```suggestion
This change removes some convenience from the logging macros, as ``std::string``s are no longer accepted as the format argument.

If you previously had code like:

.. code-block::

  std::string my_std_string = "Foo %d";
  RCLCPP_DEBUG(get_logger(), my_std_string, 5);
  
you should now replace it with:

.. code-block::

  std::string my_std_string = "Foo %d";
  RCLCPP_DEBUG(get_logger(), my_std_string.c_str(), 5);

``std::stringstream`` types are still accepted as arguments to the stream logging macros.
See `ros2/rclcpp#1442 <https://github.com/ros2/rclcpp/pull/1442>`_ for more details.
```

clalancette reviewed (changes requested):
I added some suggestions to give users more context on what they may need to change.

clalancette commented:
(also, this PR should not be merged until we land https://github.com/ros2/rclcpp/pull/1442)

audrow commented:
I've updated the `std::string` with format args case to mention that doing so will raise a `-Wformat-security` warning ([example here](https://repl.it/@audrow/Throw-warning-on-string-conversion#main.cpp)), talked about how to get around this, and added an example of using `std::string` with no format arguments.

clalancette commented on source/Releases/Release-Galactic-Geochelone.rst:
```diff
@@ -102,6 +102,68 @@ Waitable API was modified to avoid issues with the ``MultiThreadedExecutor``.
+    std::string my_std_string = "Foo";
+    RCLCPP_DEBUG(get_logger(), "%s", my_std_string.c_str());
+
+If you are using a ``std::string`` as a format string with format arguments, converting that string to a ``char *`` and using it as the format string will yield a format security warning. To avoid the security warning, we recommend you build the string manually and pass it as in with no format arguments like the previous example.
+
+If you don't care about the ``-Wformat-security`` warning and you previously had code like:
+
+.. code-block::
+
+  std::string my_std_string = "Foo %d";
+  RCLCPP_DEBUG(get_logger(), my_std_string, 5);
+
+you could replace it with:
+
+.. code-block::
+
+    std::string my_std_string = "Foo %d";
+    RCLCPP_DEBUG(get_logger(), my_std_string.c_str(), 5); // raises -Wformat-security warning
```
Honestly, I think this is somewhat redundant with the previous example.  Also, I think we should explain why it ends up with the `-Wformat-security` warning.  So I'd think about replacing this entire block with:

```
Note: if you are using a ``std::string`` as a format string with format arguments, converting that string to a ``char *`` and using it as the format string will yield a format security warning. That's because the compiler has no way at compile to introspect into the ``std::string`` to verify the arguments.  To avoid the security warning, we recommend you build the string manually and pass it in with no format arguments like the previous example.
```

audrow replied:
Good call, I think this makes it much clearer.
